Updates docs on nested scaffolds. Issue #64136

emem365 opened the pull request:
## Description

Updating description for *Nested Scaffolds*.  The doc earlier didn't make it clear if we could use multiple scaffolds for multiple routes. Rather it mislead, making users think that there should only be one Scaffold for a whole MaterialApp, which is not the real intent.

## Related Issues
One of the latest issues mentioning this is #64136

## Tests

Only documentation changed. No tests needed

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

emem365 committed:
Updates docs on nested scaffolds

HansMuller reviewed (commented):
I think this is still a little too strongly worded. Apps often use a Scaffold at the top of a route's contents - although that's not required - and they typically don't nest Scaffolds. Simple tabbed UIs based on a BottomNavigator don't need nested Scaffolds however, complex tabbed UIs that operate like a web browser can be implemented with nested scaffolds.

This article covers the options in some detail: https://medium.com/flutter/getting-to-the-bottom-of-navigation-in-flutter-b3e440b9386

emem365 commented:
I'll take a look, thanks

Aman9026 commented:
@emem365 Remove trailing spaces to pass the tests, also don't ask to refrain.

emem365 committed:
Editting docs for Nested Scaffolds

emem365 commented:
@HansMuller @Aman9026 , would be glad if you could take a look now.
I have tried to make it a little obvious that generally scaffolds should be used only once, but if the users need to, they might nest scaffolds to their choice to achieve the functionality they need.
Hope it doesn't still sound strongly worded.

liyuqian committed:
Save startup timeline (#65118)

This would help us investigate issues like
https://github.com/flutter/flutter/issues/64781

balvinderz committed:
[flutter_tools] generates version.json for web using flutter tool (#64644)

Generates version.json in web directory of project. version.json can be used in the future version of package_info which will support web. version.json has the following keys : app_name, version and build_number. This file is generated every time a user runs or builds the project.

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

engine-flutter-autoroll committed:
Roll Engine from c6121c5d515c to 3c9308faad0b (4 revisions) (#65870)

liyuqian committed:
Continue the clipBehavior breaking change (#61366)

This follows https://github.com/flutter/flutter/pull/59364 and cl/319911104

FittedBox is still default to hardEdge clip as new FittedBox is added to Google very quickly. Let's first roll other part of changes into Google first.

LongCatIsLooong committed:
Revert "Add CompositedTransformFollower.{followerAnchor, leaderAnchor} for custom anchoring (#64930)" (#65871)

This reverts commit 2fbb529d2b04ab8a3472e14c4efdee9a068ea0f4.

devoncarew committed:
fix an issue where raw json output is written to IDE clients (#65508)

fix an issues where raw json output was written to IDE clients

liyuqian committed:
Collect memory metrics (#65798)

gspencergoog committed:
Fix the `character` field of the `RawKeyEvent` to hold correct data on non-Android platforms. (#65667)

This fixes a problem where the character field of the RawKeyEvent was not being set at all for non-Android platforms.

I also updated the key maps, and corrected a problem with the Windows key map where the backquote character wasn't correctly mapped.

Commit:
[flutter_tools] make local engine integration testing easier (#65802)

engine-flutter-autoroll committed:
Roll Engine from 3c9308faad0b to 770b143a855e (2 revisions) (#65883)

engine-flutter-autoroll committed:
1cabedf8a running screenshot tests on ios-safari unit tests (flutter/engine#20963) (#65889)

dkwingsmt committed:
Expose GestureBinding.handlePointerEvent, replacing dispatchEvent as the preferred way to dispatch events (#64846)

Commit:
[flutter] elide semantic information from certain widget spans (#65857)

clocksmith committed:
Update Navigation Rail test with regression comment and cleaner size checking (#65877)

engine-flutter-autoroll committed:
Roll Engine from 1cabedf8aca8 to 1ef10b240e28 (4 revisions) (#65892)

CareF committed:
migrate animated_placeholder_perf (#64478)

godofredoc committed:
Enable mac sdk tests and remove them from cirrus. (#65896)

HansMuller committed:
Updated references to obsolete Material button classes in benchmarks/macrobenchmarks (#65900)

HansMuller committed:
Updated references to obsolete Material button classes in examples (#65904)

YeungKC committed:
Fix FormFieldState value not in sync with the onChanged value from TextFormField. (#65695)

Commit:
[manual roll] Roll Engine from 1ef10b240e28 to f84e7a019663 (12 revisions) (#65949)

HansMuller committed:
Updated references to obsolete Material button classes in microbenchmarks (#65950)

liyuqian committed:
Mark nonflaky tests as such (#65898)

MichaelRFairhurst committed:
Remove unused 'dart:async' imports. (#65568)

justinmc committed:
TextField constrained layout bug (#65966)

Fix a bug in the layout of some textfields with constrained height and visual density.

a14n committed:
migrate some cupertino files to nullsafety (#65880)

thecalamiity committed:
Added sample code to AnimatedAlign (#64698)

bkonyi committed:
Reland "Re-enable the Dart Development Service (DDS) (#64671)" (#65873)

This reverts commit 66b01c1f2905f6b4f845f7283abe3e458351fe2e.

* Add DDS ipv6 support

* Use --dds-port for DDS instead of hijacking --host-vmservice-port

Commit:
Find text containing in tests (#65072)

nero-angela committed:
[ReorderableListView] remove extra margin added after picking up the item (#65080)

dnfield committed:
Make parameter optional (#65444)

Commit:
[flutter_tools] automatically update to latest sw on install (#65784)

If a new service worker is installed, automatically update this behind the scenes. Immediately after a page refresh, the new worker and cache will be available, though the main.dart.js and others will be available sooner due to the cache busting URLS

engine-flutter-autoroll committed:
Roll Engine from f84e7a019663 to 933f811d94ab (8 revisions) (#65975)

darrenaustin committed:
TextSelectionTheme support (step 2 of 3) (#65044)

chunhtai committed:
always adds alert label for alert dialog in Android (#65973)

gaaclarke committed:
Added the machine's architecture to macos doctor results. (#65978)

Commit:
Revert "always adds alert label for alert dialog in Android (#65973)" (#66031)

This reverts commit fd0554ba89e6e2904355bc96622554b6a4ca6bf9.

renyou committed:
Revert "TextField constrained layout bug (#65966)" (#66027)

This reverts commit 10a66b19d5a7b7169602af74795c80e139907266.

Commit:
Revert "TextSelectionTheme support (step 2 of 3) (#65044)" (#66051)

This reverts commit cb92ffc7fa2a3731b98bd4dc7ddfedcc1f50ea6e.

bkonyi committed:
Roll packages to fix #66038 (#66052)

PR #65873 enabled DDS for non-web Flutter applications. This change
surfaced a bug where DDS would not send ServiceRegistered events to VM
service clients when they subscribed to the Service stream. This would
cause tools and tests that rely on these events (e.g., hot reload in
VSCode) to fail.

Fixes #66038.

HansMuller committed:
Replaced reference to obsolete FlatButton button class in SnackBar (#65988)

jmagman committed:
Inform user how to fix permissions when the observatory URL isn't found on iOS 14 (#65977)

MeneDev committed:
Fixed typo (subetting -> subsetting) (#62502)

hpoul committed:
Remove FlutterApplication from app templates. (#64742)

jmagman committed:
Check that header exists instead of content (#65961)

Commit:
Deprecate VelocityTracker default constructor and added VelocityTracker.withKind constructor (#66043)

We've gotten feedback that the VelocityTracker change was disruptive, though it did not break any of the flutter framework or customer tests. In order to make the change non-breaking, PointerDeviceKind parameter can be made optional.

Nevertheless, this parameter should be provided so that the existing touch handlers can use more accurate gestures on mouse/stylus inputs, so we can encourage this by deprecating the default constructor and pointing users towards the VelocityTracker.withKind constructor that takes a non-optional parameter

Commit:
[flutter_tools] use flutter tool handler for dwds resources and precache tool pub dependencies (#65814)

If the tool is downloaded from a precompiled snapshot, or if the backing source files in the pub cache are deleted, the dwds debugging functionality will break as the client.js file cannot be located. Instead use the PackageConfig to verify that package location, downloading if it is missing.

Override the dwds middleware to avoid Isolate.resolvePackageUri

Fixes #53644
Fixes #65475

thecalamiity committed:
Added clipBehavior to Overlay, Flow, AnimatedSize and AndroidView (#65910)

jnschulze committed:
update_dart_sdk.ps1: Ensure Start-BitsTransfer always throws an exception in case of errors (#65929)

stuartmorgan-g committed:
Add VERSIONINFO to the Windows template (#66025)

Adds a VERSIONINFO to Runner.rc in the Windows app template, populated
from the project creation metadata.

Currently the version itself is hard-coded, but it is future-proofed to
allow plumbing the actual version through at build time via preprocessor
defines.

stuartmorgan-g committed:
Update Windows system font change handling (#66021)

The engine now has an explicit API for system font changes, rather than
expecting the WM_FONTCHANGE message to be forwarded to the Flutter view.

stuartmorgan-g committed:
Consider the Windows app template stable (#66022)

Removes the template version from the Windows template; the API and
tooling boundary will now be considered stable, so there will no longer
be frequent breaking changes.

Also updates the link for adding desktop support to a project for all
three platforms to reflect the current location.

Fixes https://github.com/flutter/flutter/issues/52748

Commit:
[flutter_tools] update windows config feature (#65867)

mzdm committed:
Minor docs updates (#64966)

xu-baolin committed:
Fix the inconsistency between the local state of the input and the engine state (#65754)

younghwan committed:
Apply darkmode style (#65182)

godofredoc committed:
Enable pre/post submit hostonly tests. (#65985)

mdebbar committed:
[web] Inform the engine when read-only flag is flipped (#65499)

darrenaustin committed:
Reland ##65044 (#66061)

Turn the opt-in default for ThemeData.useTextSelectionTheme to true so that everything uses the new TextSelectionTheme by default.

DanielEdrisian committed:
Updated androidMissingSdkInstructions error message (#65964)

yjbanov committed:
Hide flutter test --platform (#65984)

engine-flutter-autoroll committed:
2abe69c60 Roll Dart SDK from 83365979ea85 to e41a6008746d (1 revision) (flutter/engine#21224) (#65987)

a14n committed:
remove non-nullability on Navigator methods (#65997)

chunhtai committed:
reland always adds alert label for alert dialog in Android (#66057)

Commit:
[flutter_tools] optimize fetch requests and remove main.dart.js bypass (#66069)

The main.dart.js bypass is not needed now that we have skipWaiting. Additionally optimize the fetch handler so that resources not in the cache skip the service worker altogether.

Fixes #66068

gusghrlrl101 committed:
Change LicensePage's loading color from scaffoldBackgroundColor to cardColor (#64639)

* wrap LicensePage's loading widget with cardColor

* Set AnimatedSwitcer's duration in LicensePage to zero

* Remove AnimatedSwitcher

* Add test code for checking color is same

dnfield committed:
Fix test file name (continous -> continuous) (#66081)

bkonyi committed:
Ensure VmService instance is disposed after failed direct connection attempt (#66123)

_attemptServiceConnection in FallbackDiscovery would fail if the root
library URI was not a package URI even if the VM service connection
attempt was successful. This resulted in a VM service connection being
left alive, causing DDS to fail its connection to the VM service.

Updated _attemptServiceConnection to ensure the VM service instance is
disposed after a non-connection related failure and to allow for root
library URIs with a file scheme.

haeseoklee committed:
Add dart-pad example code for CupertinoSliverRefreshControl (#65164)

stuartmorgan-g committed:
Remove the Windows 'flutter create' warning (#66113)

darrenaustin committed:
Add .DS_Store to ignore list for the analyzer to run on macos cleanly. (#65530)

speaking-in-code committed:
Make sure all isolates start during flutter driver tests. (#65703)

yash1200 committed:
Changed title to label in bottom_navigation_bar_test.dart (#65918)

Piinks committed:
Add versioning to gold endpoints (#66054)

xu-baolin committed:
Fix bug when updating the `divisions` and `value` of the slider at the same time (#65998)

orestesgaolin committed:
Fix mistake in the docs of RouteInformationParser (#66023)

justinmc committed:
Reland "TextField constrained layout bug (#65966)" (#66055)

jmagman committed:
Stream logging from attached debugger on iOS (#66092)

zanderso committed:
Revert "Ensure VmService instance is disposed after failed direct connection attempt (#66123)" (#66164)

This reverts commit 409e994cd03749d5a903f01cf4ec5b3ee074a2ee.

Co-authored-by: Jenn Magder <magder@google.com>

jmagman committed:
Turn off flutter_gallery__transition_perf_e2e_ios32 in devicelab (#66187)

Commit:
[flutter_tools] connect widget cache from frontend_server (#65951)

justinmc committed:
InteractiveViewer onInteractionUpdate focalPoint (#66065)

Commit:
[versions] update to latest source span and roll engine to  4b8477d11573d233e6791204191c0090f733b05d (#66136)

Piinks committed:
Add handling for empty expectation (#66142)

Commit:
[devicelab] fix manifest definition (#66277)

Rahiche committed:
add ScrollViewKeyboardDismissBehavior to CustomScrollView constructor (#66014)

ctrysbita committed:
Remove deprecated activity indicator (#66020)

Commit:
[flutter_tools] fix bug where last build id parent folder is missing (#66156)

Hixie committed:
Prevent a potential infinite loop. (#66073)

godofredoc committed:
Enable linux and win hostonly devicelab tests. (#66085)

Commit:
[flutter_tools] add EPERM to set of immediate exit errors (#66159)

yjbanov committed:
roll source_span 1.8.0-nullsafety.2 (#66290)

Commit:
[flutter_tools] declare non-transitive deps correctly (#66269)

liyuqian committed:
Mark animated_placeholder_perf__e2e nonflaky (#66284)

justinmc committed:
TextField intrinsic height layout bug (#66291)

Fixes a bug when TextField is wrapped in IntrinsicHeight and visual density is below standard.

Commit:
[flutter_tools] remove k toggle for canvaskit from web runner (#66273)

The k toggle allows switching between canvaskit and html backend at runtime. unfortunately this causes hot restart to break, since the dart_sdk modules stores state at runtime. The recommendation will be to use cavaskit via dart-defines.

dnfield committed:
Revert "Stream logging from attached debugger on iOS (#66092)" (#66359)

This reverts commit 5c8580360aa6e98c560dfe9191e0d42ad2015f9e.

dnfield committed:
Revert "Revert "Stream logging from attached debugger on iOS (#66092)" (#66359)" (#66360)

This reverts commit 53fee1be383b34416d83431df676c5546f4e36fa.

jmagman committed:
Revert "Stream logging from attached debugger on iOS (#66092)" (#66368)

This reverts commit 5c8580360aa6e98c560dfe9191e0d42ad2015f9e.

dnfield committed:
Allow new methods to be added to ui.Image for tests (#65876)

Remove fake image implementations, add createTestImage to flutter_test

ny927 committed:
Add sample code to DefaultTextStyleTransition (#64638)

Commit:
[flutter_tools] dont crash if attach is given a bad debug uri (#66358)

godofredoc committed:
Fix windows hostonly_devicelab tests. (#66195)

xu-baolin committed:
SliverList perform layout start from the initial child when there is no valid layout offset (#66262)

goderbauer committed:
Make CupertinoThemeData properties non-nullable (#66274)

satsukiu committed:
Reland fuchsia_remote_debug_protocol allows open port on remote device (#66271)

jmagman committed:
Revert "Turn off flutter_gallery__transition_perf_e2e_ios32 in devicelab (#66187)" (#66302)

goderbauer committed:
Bump dartdoc to 0.34.0 (#66048)

dnfield committed:
Mark Windows gradle_plugin_light_apk_test as flaky (#66391)

This test is failing and was recently added.

jmagman committed:
Stream logging from attached debugger on iOS (#66092) (#66390)

jmagman committed:
Revert "Stream logging from attached debugger on iOS (#66092) (#66390)" (#66397)

This reverts commit 2be4570d3a3b3a54cdbe48f5ca162d21e1a90351.

dnfield committed:
Remove last FakeImage implementation (#66379)

bkonyi committed:
Revert "Revert "Ensure VmService instance is disposed after failed direct connection attempt (#66123)" (#66164)" (#66377)

This reverts commit 3e6863ab2bc5cd33b41646ec957969860af7d518.

Commit:
[flutter_tools] map file Uri to multi-root scheme if provided (#66151)

If a file scheme and one or more roots is provided, fall back to this mapping before the direct file path if the file path cannot be turned into a package URI.

Fixes #66095

Calamity210 committed:
List tile docs (#65584)

Commit:
[flutter_tools] allow device classes to provide platform-specific interface for devFS Sync (#66266)

JuYeong0413 committed:
Fix CupertinoAlertDialog TextStyle (#64468)

liyuqian committed:
Add Overflow back with deprecation (#66305)

jmagman committed:
Listen to Debug VM stream to get Stdout logs from VMService (#66310)

jmagman committed:
Stream logging from attached debugger on iOS 13+ (#66399)

GaryQian committed:
Provide defaulting for textScaleFactor when passing to dart:ui (#66375)

engine-flutter-autoroll committed:
Roll Engine from 4b8477d11573 to c62d6652992b (44 revisions) (#66398)

Commit:
[flutter_tools] only lock if an upgrade/download will be performed (linux/macos) and output building messages to stderr (#65422)

Currently an invocation of flutter/dart will always attempt to acquire a lock. This can pose problems for tools that attempt to run multiple dart/flutter instances.

Instead update the lock logic (on Linux/macOS) so that we only attempt to acquire it if an update/snapshot needs to be performed. To avoid repeatedly performing downloads/snapshots if multiple flutter/dart invocations are fired off concurrently when an update needs to be performed, do a second check of the download/snapshot condition after the lock is released.

Additionally, moves all of the building/debug output to stderr on both the bash and batch scripts. This allows machine mode consumption of the tool to ignore needing to parse/handle the rebuild messages.

jmagman committed:
Revert "[flutter_tools] map file Uri to multi-root scheme if provided (#66151)" (#66403)

This reverts commit af6ba86728c134bdb80c7a250fc0184078b36d60.

LongCatIsLooong committed:
Positioning IME bars on iOS (#61981)

engine-flutter-autoroll committed:
339528342 Roll Skia from c61c895393ea to 2b469ebd0627 (9 revisions) (flutter/engine#21340) (#66400)

hamdikahloun committed:
Fix Semi Hidden helpText in showDatePicker (#65010)

Wrapping the spacing SizedBox in Flexible.

CareF committed:
migrate backdrop_filter_perf (#64482)

engine-flutter-autoroll committed:
57ee3d4b3 Roll Skia from 2b469ebd0627 to 3eb813e0cc13 (2 revisions) (flutter/engine#21343) (#66410)

a14n committed:
enable lint unnecessary_nullable_for_final_variable_declarations (#66387)

a14n committed:
migrate some cupertino files to nullsafety (#66024)

* migrate some cupertino files to nullsafety

* address review comments

* address review comments

a14n committed:
fix nullability issues (#66418)

engine-flutter-autoroll committed:
Roll Engine from 57ee3d4b3210 to 3e7ca5b86904 (2 revisions) (#66440)

Commit:
[flutter_tools] reland: map file URIs to a multiroot scheme (#66405)

If a file scheme and one or more roots is provided, fall back to this mapping before the direct file path if the file path cannot be turned into a package URI.

Use URI representation so that the transformation is resilient to the org-dartlang-app scheme used by the web builds.

Fixes #66095
Fixes #66404

Commit:
Update localizations (#65993)

thecalamiity committed:
Make Dismissible's HitTestBehavior an argument (#64379)

Commit:
Expose enable-experiment in Flutter drive (#66311)

Commit:
[flutter_tools] do not add events to closed sink in throttle transform (#66468)

The throttle duration could delay past the point where the destination sink was closed. Check if it is closed before adding an event. Fixes a crash on dev: StateError: Bad State: Stream is already closed.

thecalamiity committed:
Updated README.md file of the hello_world example (#64766)

jakemac53 committed:
update to the latest null safe packages (#66384)

Updates all null safe dependencies to versions that allow 2.10 stable and 2.11 dev releases.

Also updates flutter_goldens and flutter_goldens_client to allow 2.11 dev.

gspencergoog committed:
Change the default visual density to be adaptive based on platform. (#66370)

Based on feedback from various desktop developers, and to be consistent between the defaults and the sample code, this PR switches the default for visual density in desktop themes to be compact instead of standard.

It also removes the same setting from the sample code generated by "flutter create" because it is no longer needed now that it is the default.

Commit:
[devicelab] increase timeout for cull bench (#66481)

jmagman committed:
Check git commands in Flutter version check test (#66406)

Commit:
[flutter_tools] flush UI thread tasks before finishing hot restart (#66417)

Improve our ability to measure the performance of hot restart. Also flush all UI thread tasks before calling the hot restart "complete" in benchmark mode.

If this is successful at stabilizing the benchmarks, this should be enabled for "normal" hot restart so that we can accurately track execution time.

a14n committed:
migrate some cupertino files to nullsafety (#66424)

* migrate some cupertino files to nullsafety

* address review comments

engine-flutter-autoroll committed:
53c0d0c02 Roll Skia from 3eb813e0cc13 to 18f4b1c7e31a (4 revisions) (flutter/engine#21351) (#66462)

engine-flutter-autoroll committed:
Roll Engine from 53c0d0c02af0 to dd35b5b65917 (10 revisions) (#66497)

Commit:
[flutter_tools] try deleting the web cache directory before copying new web sdk (#66461)

Delete the destination directory for the web sdk before copying the new contents. This might fix an error where the OS refuses to overwrite the files on dev

godofredoc committed:
Fix gradle_plugin_light_apk test. (#66496)

Commit:
[flutter_tools] enable single widget reload optimization by default on master (#66507)

This should be a mostly non breaking change for hot reload workflows that improves performance, but if you see anything odd with hot reload behavior on master please file a bug. This feature can also be disabled without any changes to the SDK with flutter config --no-single-widget-reload-optimization

Commit:
Remove skipped tool test, remove integration shard from cirrus (#66503)

shihaohong commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -964,13 +964,18 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
-/// [TabBarView], you might be tempted to make each tab bar view a
-/// scaffold with a differently titled AppBar. It would be better to add a
-/// listener to the [TabController] that updates the AppBar.
+/// The Scaffold was designed to be a single top level container for
+/// most of the content in a route(although, not necessarly the first widget).
+/// It is recommended that you use Scaffold once per route(adhering to
+/// material guidelines) for most of the simple cases, unless you really
+/// need to nest them to achieve the functionality you need.
+///
+/// You'll find that it's typically not necessary to nest scaffolds. For
+/// example, in a tabbed UI, where the [bottomNavigationBar] is a [TabBar]
+/// and the body is a [TabBarView], you might be tempted to make each tab bar
+/// view a scaffold with a differently titled AppBar. Rather, it would be
+/// better to add a listener to the [TabController] that updates the
+/// AppBar.
```
A few comments and my suggestion:
- "First widget" is vague here -- Did you mean the top-level widget?
- I would avoid using words like "recommend" or "refrain", since a user has the liberty to decide what they want to do with their apps. We should opt to describe what the widget does and what outcome it'll help the developer achieve.

```suggestion
/// The Scaffold is designed to be a top level container for
/// a [MaterialApp]. This means that adding a Scaffold
/// to each route on a Material app will provide the app with
/// Material's basic visual layout structure.
///
/// It is typically not necessary to nest Scaffolds. For example, in a 
/// tabbed UI, where the [bottomNavigationBar] is a [TabBar]
/// and the body is a [TabBarView], you might be tempted to make each tab bar
/// view a scaffold with a differently titled AppBar. Rather, it would be
/// better to add a listener to the [TabController] that updates the
/// AppBar.
```

emem365 replied:
Sure. These make a lot more sense than before, and also subtly bring across the fact that we can add a scaffold to each route. Committing these changes

yash1200 committed:
Changed title to label (#66186)

Commit:
[flutter_tools] register service worker after first frame event (#66082)

Registering the service worker immediately after the documented has loaded may cause SW initialization to compete with framework initialization. It was recommended to us that we defer the service worker setup until after the framework is done with setup, which should be sometime after the first frame.

To implement this, I augmented the binding setup to dispatch an event on the document after the binding has initialized. I don't see any obvious risks with this setup.

Fixes #66066

gspencergoog committed:
Change emoji in About dialog to be a divider (#66409)

pq committed:
disable pub lints (#66522)

* disable pub lints

* + comments

willlarche committed:
[Icons] Update icon version to point to file that iOS will prefer. (#66524)

thecalamiity committed:
Fixes typos in showDialog documentation (#66213)

* Fixes typos in showDialog documentation

goderbauer committed:
Allow futures that resolve to null in Future/StreamBuilder (#66506)

darrenaustin committed:
Complete the migration to the TextSelectionTheme by deprecating the opt-in flag and text selection properties in ThemeData. (#66482)

willlarche committed:
[Docs] [Material] Fix Icons api docs (#66508)

* [Docs] [Icons] Updating dart doc styles to recognize more styles of icons. All the 4 icon styles we support have to be pulled from 4 individual web fonts when we show the html page for api docs.

* Better documentation and that special awful case.

* Couple more places to remove.

Commit:
[flutter_tools] bypass pubspec yaml content check when running pubdependencies (#66559)

Otherwise the heuristics we have may decide to skip running pub. We should consider reducing the number of dimensions of caching here to a single check optimized for the run/test case.

engine-flutter-autoroll committed:
Roll Engine from dd35b5b65917 to b0fb2c8a988f (2 revisions) (#66505)

* b49de93b3 Create an ImageHandle wrapper (flutter/engine#21057)

* b0fb2c8a9 Roll Skia from 7b97b3cb2bd0 to 59b2a92c96ba (4 revisions) (flutter/engine#21365)

DetectiveAI4869 committed:
[flutter_tools] Add channel order aware version_test (#62417)

liyuqian committed:
Default measureCpuGpu to true (#66386)

a14n committed:
migrate cupertino to nullsafety (#66493)

liyuqian committed:
Revert "Default measureCpuGpu to true (#66386)" (#66594)

This reverts commit 48ce83a833e7cb3c27e73e238f0d565470827139.

engine-flutter-autoroll committed:
Roll Engine from b0fb2c8a988f to e56080ab6777 (17 revisions) (#66583)

jmagman committed:
Do not cleanup Gradle in non-Android integration tests (#66487)

Commit:
[flutter_tools] fix calling debugToggleBrightness on release mode (#66401)

ensure that the terminal handler checks if the service protocol is enabled before calling debug toggle brightness. Also removes globals from TerminalHander and test cases.

Fixes #65477

engine-flutter-autoroll committed:
Roll Engine from e56080ab6777 to 435db234e4d3 (12 revisions) (#66623)

LongCatIsLooong committed:
Length formatter minor fix (#66567)

androiddevnotes committed:
Fix typos (#66029)

engine-flutter-autoroll committed:
Roll Engine from 435db234e4d3 to 62b5a53b10a1 (2 revisions) (#66630)

engine-flutter-autoroll committed:
3a73d073c Roll Skia from 5227335b0add to 1748c6a3b8c8 (1 revision) (flutter/engine#21402) (#66638)

HansMuller committed:
Replaced obsolete use of FlatButton with TextButton (#66597)

HansMuller committed:
Replaced obsolete use of FlatButton with TextButton in scroll_activity_test (#66602)

johnsonmh committed:
[Material] Remove opacity from dark theme overlay check (#66596)

* Remove opacity from check and add test

HansMuller committed:
Replaced use of obsolete RaisedButton with ElevatedButton in widget_inspector_test (#66603)

chinmoy12c committed:
Changed TickerProviderStateMixin to SingleTickerProviderStateMixin in ScrollbarState (#66653)

emem365 committed:
Nested Scaffolds - Suggested Changes

Co-authored-by: Shi-Hao Hong <shihaohong@google.com>

emem365 committed:
Nested Scaffolds - Suggested Changes

Co-authored-by: Shi-Hao Hong <shihaohong@google.com>

emem365 committed:
Nested Scaffolds -  Suggested Changes

emem365 committed:
Nested Scaffolds -  Suggested Changes

gaaclarke committed:
Added timeout for closing devfs sync http connections. (#66152)

liyuqian committed:
Roll gallery to the newest version (#66584)

This shouldn't break our tests as
https://github.com/flutter/gallery/issues/275 is closed.

jmagman committed:
Remove Try accepting the local network permissions popup warning (#66621)

godofredoc committed:
Make Gradle plugin light apk test blocking again. (#66587)

liyuqian committed:
Revert "Roll gallery to the newest version (#66584)" (#66666)

This reverts commit 2ae09caace4545d9fbbfee281196a97f6db556ce.

jason-simmons committed:
Add a compressionState value to HttpResponse mocks (#66679)

The Dart SDK now requires non-null enum values if the enum is used in an
exhaustive switch statement on a non-nullable type.

See https://github.com/flutter/flutter/issues/66674

godofredoc committed:
Add Firebase tests to flutter dashboard. (#66677)

This PR also enables firebase tests on presubmit and removes code that
won't be used anymore from test runner.

Bug:
  https://github.com/flutter/infra/pull/225

Commit:
[flutter_tools] ensure ErrorHandlingFileSystem wraps current directory (#66680)

The lack of current directory wrapping was letting some of the already handled errors through

Fixes #66675

NebiSarigul committed:
Update .gitignore.tmpl (#66382)

mdebbar committed:
[web] Change the web server to support path url strategy (#66606)

Commit:
Teach the flutter tool how to find android output files if the flavor contains uppercase letters (#66687)

This changes the name of the generated output files on android in case a flavor with uppercase letters is used.
Previously, the lowercased flavor name would be used for the apk/aab file. Now, the flavor name is used as-is.

justinmc committed:
Update docs about complex character support (#66582)

Simple docs fix for characters.

b1acKr0se committed:
Fix last month not being displayed if last date is selected in DateRangePicker (#66542)

willlarche committed:
[Icons][iOS] Pointing to version of material icon font that includes more metadata in the xml. (#66684)

Commit:
[flutter_tools] pass existsSync through error handling io (#66704)

Crash reporting shows at least one instance of EACCES during an existsSync cache call. Add this to the list of error handling io methods. Did not add the async exists method since we lint against its usage.

Commit:
[flutter_tools] dont mention git clone of flutter in run message (#66691)

This message is displayed if we do not find a pubspec.yaml

99% of the time this additional advice is going to have nothing to do with the actual error. Just remove it because it will cause more confusion than it could clear up

a14n committed:
apply upcoming lint cast_nullable_to_non_nullable (#66455)

* apply upcoming lint cast_nullable_to_non_nullable

* make RenderObjectElement.renderObject non-nullable

a14n committed:
fix the tree (#66783)

Commit:
Improve consistency of top-level help text (#66748)

goderbauer committed:
Page-subclasses to take children instead of builder (#66694)

xu-baolin committed:
Improve the behavior of Scrollable.ensureVisible when Scrollable nested (#65226)

JoseAlba committed:
Divider with subheader example update (#65944)

* Divider with subheader example update

* fixed issues

* Addressed comments

* fixed issues

shihaohong commented:
Could you rebase this branch with master? It seems like some customer test checks are failing for reasons unrelated to your PR.

shihaohong reviewed (approved):
LGTM. The failing check seems unrelated to the PR and may simply require a rebase from `master`.

Hixie committed:
Minor tool text typo (#66801)

pedromassango committed:
Add back the autovalidate class property (#66267)

* Add back autovalidate property

* Add autovalidate property back (include tests)

* Fix typos

* One more test :)

Commit:
[gallery] add linux and windows files to gallery example (#66736)

Check in linux and windows platform code now that they are stable, so that we could use in devicelab in the future. Removed the ICO from the windows example to avoid analysis check, and since it won't be important for benchmarking or UI tests

jmagman committed:
Force plugins to inherit minimum iOS version from Flutter app (#66590)

liyuqian committed:
Reland "Default measureCpuGpu to true (#66386)" (#66604)

This reverts commit 2d9e500523a71f2da57f7eea9654ffa99f001a0d.

In addition, this fixes broken Android tests of the original PR.

liyuqian committed:
Mark nonflaky tests as such (#66595)

engine-flutter-autoroll committed:
Roll Engine from 3a73d073c8be to 8d165faca1ce (45 revisions) (#66840)

* a1db2b304 Roll Fuchsia Linux SDK from NeYDIjo58... to BFLXvCMVi... (flutter/engine#21403)

* faeff0a89 Roll Dart SDK from eb8e6232da02 to 13b3f2d7b6ea (3 revisions) (flutter/engine#21407)

* 7dfcde1ff [Fix] Replaces call to deprecated method Name.name. (flutter/engine#21241)

* 48ab5d1d1 Roll Skia from 1748c6a3b8c8 to 3b88c0772e89 (1 revision) (flutter/engine#21410)

* aa8d5d4b9 Avoid sending a 0 DPR to framework (flutter/engine#21389)

* 67fdd7ede Embedder API Support for display settings (flutter/engine#21355)

* 1fc87c038 Roll Skia from 3b88c0772e89 to d7ab45027877 (1 revision) (flutter/engine#21411)

* de5f2b48f Revert "Revert "Adds fuchsia node roles to accessibility bridge updates. (#20385)" (#20936)" (flutter/engine#21367)

* c9b40c67f Remove ASCII art from mDNS error log (flutter/engine#21397)

* fed05315a Roll Fuchsia Mac SDK from xnB_uJM8T... to _e0onA6gY... (flutter/engine#21414)

* d877b830c [web] enable ios safari screenshot tests (flutter/engine#21226)

* a8f52e52c Roll Skia from d7ab45027877 to aeae3a58e3da (6 revisions) (flutter/engine#21415)

* 827560967 Roll Skia from aeae3a58e3da to 7bd60430299f (1 revision) (flutter/engine#21417)

* 65c1122bb Roll Dart SDK from 13b3f2d7b6ea to 4fb134a228c7 (2 revisions) (flutter/engine#21419)

* d735f2c59 Roll Fuchsia Linux SDK from BFLXvCMVi... to XcAUWQUZm... (flutter/engine#21420)

* f1961e520 Roll Skia from 7bd60430299f to 68861e391313 (14 revisions) (flutter/engine#21421)

* 08cf7256d Fix getNextFrame (flutter/engine#21422)

* db9991279 Support dragging native platform views (flutter/engine#21396)

* df83e8f17 Roll Skia from 68861e391313 to a05d27b170ee (1 revision) (flutter/engine#21424)

* fc7d0fc90 [web] Respond with null for unimplemented method channels (flutter/engine#21423)

* 02b85674b Roll Skia from a05d27b170ee to 5e1545fa00c8 (1 revision) (flutter/engine#21425)

* 6e54e68ba Roll Dart SDK from 4fb134a228c7 to db7eb2549480 (1 revision) (flutter/engine#21426)

* b0cd7d1ad Roll Dart SDK from db7eb2549480 to 200e8da5072a (1 revision) (flutter/engine#21427)

* 0de5c0cfe Roll Fuchsia Linux SDK from XcAUWQUZm... to 0nW5DAxcC... (flutter/engine#21430)

* 854943d5c [macOS] Set the display refresh rate (flutter/engine#21095)

* 11aecf4df Roll Skia from 5e1545fa00c8 to 766eeb2ac325 (1 revision) (flutter/engine#21431)

* b8e2b3f89 Roll Skia from 766eeb2ac325 to 5648572f4a94 (1 revision) (flutter/engine#21433)

* 33d0bbbc0 Roll Fuchsia Mac SDK from _e0onA6gY... to SUSVNJcX5... (flutter/engine#21434)

* 51049d258 Roll Skia from 5648572f4a94 to eabce08bb2f2 (1 revision) (flutter/engine#21435)

* a491533de Roll Dart SDK from 200e8da5072a to c938793e2d6f (1 revision) (flutter/engine#21437)

* 06398b87e Roll Fuchsia Linux SDK from 0nW5DAxcC... to xdxm8rU8b... (flutter/engine#21439)

* 4282bbc4b Roll Dart SDK from c938793e2d6f to fe83360d3a7c (1 revision) (flutter/engine#21440)

* eba7b8e60 Roll Fuchsia Mac SDK from SUSVNJcX5... to v5Ko06GkT... (flutter/engine#21441)

* 249bcf7a5 Roll Fuchsia Linux SDK from xdxm8rU8b... to ej-CkfSra... (flutter/engine#21443)

* 4422ede07 Roll Fuchsia Mac SDK from v5Ko06GkT... to k_lSjZxIH... (flutter/engine#21445)

* 35fa4bbdb Roll Dart SDK from fe83360d3a7c to 44e4f3958019 (1 revision) (flutter/engine#21446)

* a9b5b133f Roll Fuchsia Linux SDK from ej-CkfSra... to HNNs4gfuM... (flutter/engine#21447)

* 4f7ff211b Roll Skia from eabce08bb2f2 to ad6aeace6eee (2 revisions) (flutter/engine#21448)

* f72613d98 Roll Dart SDK from 44e4f3958019 to e2a4eaba73b8 (1 revision) (flutter/engine#21451)

* 2917a6501 [ios] Remove unused is_valid_ from IOS Metal Context (flutter/engine#21432)

* b5916741d Roll Dart SDK from e2a4eaba73b8 to 13deada5b267 (1 revision) (flutter/engine#21453)

* a07e0e095 Roll Fuchsia Mac SDK from k_lSjZxIH... to qyoO7f9Sk... (flutter/engine#21454)

* cf1fbf271 Apply dpr transform to fuchsia accessibility bridge (flutter/engine#21364)

* 9db9a57a2 Revert "Apply dpr transform to fuchsia accessibility bridge (#21364)" (flutter/engine#21458)

* 8d165faca Revert multiple display support for embedder API (flutter/engine#21456)

Kavantix committed:
Sliver padding overlap fix (#65323)

engine-flutter-autoroll committed:
67b1219d8 Roll Skia from ad6aeace6eee to 6a189f23af5e (6 revisions) (flutter/engine#21455) (#66841)

Commit:
[flutter_tools] add a mechanism to turn off immediate tool exit (#66787)

Instead of always exiting the tool, provide a mechanism to turn off this behavior for non-critical functionality like configuration and analytics settings.

Fixes #66786 Fixes #4674

a14n committed:
migration of material files to nullsafety (#66633)

engine-flutter-autoroll committed:
Roll Engine from 67b1219d8350 to 96dd918f2d4e (3 revisions) (#66848)

Commit:
[desktop] default to shrink wrap on desktop platforms (#66754)

Commit:
Update flutter_tools README.md to document need for FLUTTER_ROOT (#66854)

engine-flutter-autoroll committed:
Roll Engine from 96dd918f2d4e to 77701d303e22 (2 revisions) (#66855)

LongCatIsLooong committed:
fix _getPixelPerfectCursorOffset logic when infinite (#66845)

engine-flutter-autoroll committed:
Roll Engine from 77701d303e22 to 438685dad025 (7 revisions) (#66868)

liyuqian committed:
Roll gallery to the newest version (#66693)

The previously broken test should be fixed by
https://github.com/flutter/gallery/pull/331

liyuqian committed:
Default FittedBox's clipBehavior to none (#66700)

This is a continuation of https://github.com/flutter/flutter/pull/61366
after cl/333620714

See also https://flutter.dev/go/clip-behavior

engine-flutter-autoroll committed:
Roll Engine from 438685dad025 to 1c462e28a2e2 (5 revisions) (#66892)

Commit:
[flutter_tools] handle missing zip/unzip argument errors (#66685)

it is possible for users that download the prebuilt SDKs to run the tool without unzip installed. Rather than crashing with an unclear argument error, exit the tool with a message on how to install that mirrors the message in the update_dart_sdk.sh script.

Also applied to zip for completeness

justinmc committed:
GlobalKey docs improvement (#66834)

Improve GlobalKey docs by mentioning a pitfall that some users had where the instantiated it in the build method.

engine-flutter-autoroll committed:
Roll Engine from 1c462e28a2e2 to 414805d1d824 (2 revisions) (#66894)

abrush21 committed:
Adds getter/setter for slider semantics flag. (#65766)

bkonyi committed:
Roll package:dds to 1.4.0 and update error handling (#66836)

engine-flutter-autoroll committed:
5a7336ef2 Roll Dart SDK from 4c0245356649 to 4ea46a8b10a1 (1 revision) (flutter/engine#21482) (#66913)

Commit:
Add the ability to inject a bootstrap script (#66897)

scheglov committed:
Move assert(s) that reference 'this' to the constructor bodies. (#66914)

engine-flutter-autoroll committed:
Roll Engine from 5a7336ef2637 to a6a6fd163b99 (3 revisions) (#66924)

Commit:
[flutter_tools] do not crash if chrome preference save fails (#66705)

Crash reporting shows at least one occurrence of this due to a windows file lock. That could happen if there is another running chrome instance, or perhaps a virus scanner is running. Print out the error and do not crash.

a14n committed:
migrate some material files to nullsafty (#66858)

* migrate some material files to nullsafty

* address review comments

CareF committed:
cleanup (#66743)

DanielEdrisian committed:
Allow modifying barrier color and barrier dismissible for Cupertino Modal Popup (#66692)

engine-flutter-autoroll committed:
Roll Engine from a6a6fd163b99 to 326827883429 (7 revisions) (#66938)

dnfield committed:
Revert "Improve the behavior of Scrollable.ensureVisible when Scrollable nested (#65226)" (#66918)

This reverts commit 73b6398932b3142a6e2f479327c50c5aa613a680.

Commit:
[flutter_tools] do not error flutter doctor on missing AS/intellij plugins (#66782)

In cases where the Intellij/AS plugins are not located, display links to where they can be downloaded but do not surface an error. This should generally reduce confusion about whether the plugins are required for every installed IDE. For example, frequently users may only install AS so that they can install the Android SDK - or they may have multiple copies of Intellij installed.

For example: #66762

engine-flutter-autoroll committed:
fff415d51 Roll Fuchsia Mac SDK from iiCa4hab1... to tuXjGvikz... (flutter/engine#21492) (#66939)

Commit:
remove --trace-startup (#66568)

CareF committed:
fix smoothness (#66611)

jmagman committed:
Replace MockFile with memory file system files (#66941)

Commit:
[flutter_tools] do not require a dependency on devtools server (#66842)

Re-arrange the implementation of the devtools launcher so that google3 is not required to depend on any devtools packages. Also renames the build_runner folders to isolated to better clarify their intention.

engine-flutter-autoroll committed:
d18c3ea75 Roll Skia from 0b0fb4d50b75 to 9ecb3abfdfe8 (12 revisions) (flutter/engine#21500) (#66942)

emem365 committed:
Merge remote-tracking branch 'origin/nested-scaffold-docs-edit' into nested-scaffold-docs-edit

emem365 commented:
okay so, This just happened as a result of me pushing after rebasing the branch with master. I dont know if this is how this works, so if this isnt how this was supposed to work I can just create a new PR with a clean branch from master. Please let me know. @shihaohong @HansMuller

pedromassango commented:
Hi @emem365 
You can close this one then.
